src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs

```diff
@@ -1577,7 +1586,7 @@ ICollection IDictionary.Values
                 if (key == null) ThrowKeyNullException();
 
                 if (!(key is TKey)) throw new ArgumentException(SR.ConcurrentDictionary_TypeOfKeyIncorrect);
-                if (!(value is TValue)) throw new ArgumentException(SR.ConcurrentDictionary_TypeOfValueIncorrect);
+                ThrowIfNullAndNullsAreIllegal(value);
```

Comment:
Seems like the fix should be to go in the other direction: rather than adding this ThrowIfNullAndNullsAreIllegal thing, undo all of that and instead just wrap this indexer setting with a try/catch like that used in IDictionary.Add.  As above, this change is turning something that was an ArgumentException into an InvalidCastException, which is not something we want to do.

Reply:
I do not understand. `ThrowIfNullAndNullsAreIllegal` method throws `ArgumentException` when `value` is `null`. There is no `InvalidCastException`. The implementation was taken from [.NET framework](https://github.com/dotnet/coreclr/blob/38cf93013c1dd1efc7137a6f4930cab7cc653411/src/mscorlib/shared/System/Collections/Generic/Dictionary.cs#L831-L838). And this is the purpose of this PR. To throw `ArgumentException` when `value` is `null` and `nulls` are prohibited (`TValue` is value type).

Reply:
Let's say TValue is Foo, and someone passes in Bar.  Previously that would have thrown an ArgumentException.  Now it'll pass the checks in ThrowIfNullAndNullsAreIllegal (it's not null), but then it'll fail with a cast exception on the next line when it's cast to Foo.

Reply:
I returned back the original type check with added condition to not be null, but left there `ThrowIfNullAndNullsAreIllegal`. The type check is for the reason you have mentioned, the other is for the addressing the `null` issue in #26447.

Just there are diffrent checks for type of the value. There is `try-catch` in `IDictionary.Add` and `if` in the indexer. Maybe it would be better to have the same type check, but I do not know, which one is better to use (from the performace view for example). I would say that it is the `if`.